### ISSUE TITLE
Missing Gradient in Fast_Path Mode?

### ISSUE THREAD

**Issue by Leopold2333** [state-spaces/mamba#342]
Very breliant work! But I got such an issue: are the gradients passed properly with `use_fast_path=True`?
https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/mamba_ssm/ops/selective_scan_interface.py#L259

Here is the code snippet in `MambaInnerFn`, and the gradient of `z` is calculated by selective_scan.bwd():
```python
dconv1d_out, ddelta, dA, dB, dC, dD, ddelta_bias, dz, out_z = selective_scan_cuda.bwd(
    conv1d_out, delta, A, B, C, D, z, delta_bias, dout_y, scan_intermediates, out, dz,
    ctx.delta_softplus,
    True  # option to recompute out_z
)
```
However, after this, dz is never used, the final gradient of variable `xz` is the `torch.empty_like(xz)` directly? Is the gradient of `z` properly passed backward?

When `use_fast_path=False`, all things seems alright, because in `SelectiveScanFn`, the gradients are passed like this:
```python
du, ddelta, dA, dB, dC, dD, ddelta_bias, *rest = selective_scan_cuda.bwd(
    u, delta, A, B, C, D, z, delta_bias, dout, x, out, None, ctx.delta_softplus,
    False  # option to recompute out_z, not used here
)
dz = rest[0] if ctx.has_z else None
dB = dB.squeeze(1) if getattr(ctx, "squeeze_B", False) else dB
dC = dC.squeeze(1) if getattr(ctx, "squeeze_C", False) else dC
return (du, ddelta, dA, dB, dC,
        dD if D is not None else None,
        dz,
        ddelta_bias if delta_bias is not None else None,
        None,
        None)
```
`dz` goes bacward properly.

**Comment by tridao** [comment]
Have you tried? The test for dz is here: https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/tests/ops/test_selective_scan.py#L131

**Comment by tridao** [comment]
Alternatively you can put a breakpoint and step through to see that dxz is being updated.

**Comment by Leopold2333** [comment]
> Alternatively you can put a breakpoint and step through to see that dxz is being updated.

 I have made a test, setting a breakpoint in the `backward` function in [selective_scan_interface.py](https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/mamba_ssm/ops/selective_scan_interface.py).

There are three parts referring to gradients of `dxz`, `dx` and `dz`:
https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/mamba_ssm/ops/selective_scan_interface.py#L259-L260
https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/mamba_ssm/ops/selective_scan_interface.py#L263-L267
https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/mamba_ssm/ops/selective_scan_interface.py#L299-L301

I set a breakpoint on line 304. I use `d_model=128` and find that `dxz[:, 128:, :]≠dz`, while `dx` remains unchanged after `causal_conv1d_cuda.causal_conv1d_bwd()` (so that `dxz[:, :128, :]=dx`. I'm not familiar with the mechanism of causal conv-1d, well, I have also checked this issue in [causal_conv1d.cpp](https://github.com/Dao-AILab/causal-conv1d/blob/main/csrc/causal_conv1d.cpp), `dx` is not changed). 

The above testings are the reasons why I raise this issue. What I want to figure out is whether `dz`, the gradient of `z`, is properly pushed back to `dxz`.

**Comment by tridao** [comment]
Is this test failing? https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/tests/ops/test_selective_scan.py#L225

You can try `dxz = torch.ones_like(xz)` to check that after `selective_scan_cuda.bwd` the 2nd half of dxz is being changed.

**Comment by Leopold2333** [comment]
> Is this test failing?
> 
> https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/tests/ops/test_selective_scan.py#L225
> 
> You can try `dxz = torch.ones_like(xz)` to check that after `selective_scan_cuda.bwd` the 2nd half of dxz is being changed (i.e. no longer 1).

Oh my god, it is not 1~ And `dz` has the same value of the last half of `dxz` after `selective_scan_cuda.bwd`.

But why? That makes me confused, does `bwd` changes the data at the specified address in memory straightly?

The test is ok. I use `pytest -s test_selective_scan.py` directly to check `test_mamba_inner_fn`, here is one of the results:
```bash
mamba/tests/ops/test_selective_scan.py::test_mamba_inner_fn[True-True-128-itype0-wtype1] Output max diff: 0.0
Output mean diff: 0.0
dxz max diff: 8192.0
dA max diff: 1310720.0
dD max diff: 0.000518798828125
ddelta_bias max diff: 2048.0
dout_proj_weight max diff: 384.0
ddelta_proj_weight max diff: 98304.0
dx_proj_weight max diff: 32768.0
dconv1d_weight max diff: 24576.0
dconv1d_bias max diff: 49152.0
PASSED
```

**Comment by Hprairie** [comment]
Backward should be accumulating the gradients on dxz. Here is the chunk of cuda code on what is happening. I have put comments in to explain what is happening.

The gradients of `dzx` can then be broken down into the `dx` and `dz` gradients with the following equation.

$$ dx = dzx \cdot silu(z) $$

$$ dz = dzx \cdot (z') \cdot out $$

Here $z`$ is the derivate of the silu function, but I was too lazy to put it in.

```cuda
if constexpr (kHasZ) {
           // We load in z out and dz
            input_t *z = reinterpret_cast<input_t *>(params.z_ptr) + batch_id * params.z_batch_stride
                + dim_id * params.z_d_stride + chunk * kChunkSize;
            input_t *out = reinterpret_cast<input_t *>(params.out_ptr) + batch_id * params.out_batch_stride
                + dim_id * params.out_d_stride + chunk * kChunkSize;
            input_t *dz = reinterpret_cast<input_t *>(params.dz_ptr) + batch_id * params.dz_batch_stride
                + dim_id * params.dz_d_stride + chunk * kChunkSize;
            input_t z_vals[kNItems], out_vals[kNItems];
            __syncthreads();
            load_input<Ktraits>(z, z_vals, smem_load, params.seqlen - chunk * kChunkSize);
            __syncthreads();
            load_input<Ktraits>(out, out_vals, smem_load, params.seqlen - chunk * kChunkSize);
            float dz_vals[kNItems], z_silu_vals[kNItems];
            // We calculate the gradients for each value described with the equations above
            #pragma unroll
            for (int i = 0; i < kNItems; ++i) {
                float z_val = z_vals[i];
                float z_sigmoid_val = 1.0f / (1.0f + expf(-z_val));
                z_silu_vals[i] = z_val * z_sigmoid_val;
                dz_vals[i] = dout_vals[i] * float(out_vals[i]) * z_sigmoid_val
                             * (1.0f + z_val * (1.0f - z_sigmoid_val));
                dout_vals[i] *= z_silu_vals[i];
            }
            // Store store the output directly back into memory at dz
            __syncthreads();
            store_output<Ktraits>(dz, dz_vals, smem_store, params.seqlen - chunk * kChunkSize);
            if (params.out_z_ptr != nullptr) {  // Recompute and store out_z
                float out_z_vals[kNItems];
                #pragma unroll
                for (int i = 0; i < kNItems; ++i) { out_z_vals[i] = float(out_vals[i]) * z_silu_vals[i]; }
                // if (blockIdx.x == 0 && blockIdx.y == 0 && threadIdx.x == 0) {
                    // printf("out_val=%f, z_silu_val = %f, out_z_val = %f\n", float(out_vals[0]), z_silu_vals[0], out_z_vals[0]);
                // }
                input_t *out_z = reinterpret_cast<input_t *>(params.out_z_ptr) + batch_id * params.out_z_batch_stride
                    + dim_id * params.out_z_d_stride + chunk * kChunkSize;
                __syncthreads();
                store_output<Ktraits>(out_z, out_z_vals, smem_store, params.seqlen - chunk * kChunkSize);
            }
        }
```

The storing will store the output directly in memory at that pointer. Since we first chunked `dz` and `dx` from `dzx` the memory pointers should still be pointing at the same tensor, i.e pytorch won't create two new tensors, but instead will point to two different positions in memory.

Thus it should update `dzx` with the values of `dx` and `dz`.

I am slightly confused about the testing though. The difference between those values is incredibly large, any clarification on that would be nice.

**Comment by tridao** [comment]
> > Is this test failing?
> > https://github.com/state-spaces/mamba/blob/a07faffa36a7b89e754b5de972418475bcdd77b6/tests/ops/test_selective_scan.py#L225
> > 
> > You can try `dxz = torch.ones_like(xz)` to check that after `selective_scan_cuda.bwd` the 2nd half of dxz is being changed (i.e. no longer 1).
> 
> Oh my god, it is not 1~ And `dz` has the same value of the last half of `dxz` after `selective_scan_cuda.bwd`.
> 
> But why? That makes me confused, does `bwd` changes the data at the specified address in memory straightly?
> 
> The test is ok. I use `pytest -s test_selective_scan.py` directly to check `test_mamba_inner_fn`, here is one of the results:
> 
> ```shell
> mamba/tests/ops/test_selective_scan.py::test_mamba_inner_fn[True-True-128-itype0-wtype1] Output max diff: 0.0
> Output mean diff: 0.0
> dxz max diff: 8192.0
> dA max diff: 1310720.0
> dD max diff: 0.000518798828125
> ddelta_bias max diff: 2048.0
> dout_proj_weight max diff: 384.0
> ddelta_proj_weight max diff: 98304.0
> dx_proj_weight max diff: 32768.0
> dconv1d_weight max diff: 24576.0
> dconv1d_bias max diff: 49152.0
> PASSED
> ```

dz is just a view into the 2nd half of dxz. Any change in dz would change dxz.
Example:
```
import torch
a = torch.zeros(4)
b = a[2:]
b[:] = 1
print(a)  # tensor([0., 0., 1., 1.])

**Comment by Hprairie** [comment]
Is there a reason why the max difference is so large, I can see that all the asserts are commented out, but am confused why this is stable if all the gradients seem messed up.

**Comment by tridao** [comment]
I think the test doesn't have the right initialization (e.g. for A, delta) so the outputs are actually extremely large, and the relative diff is still quite small. I haven't had time to revisit.

**Comment by Leopold2333** [comment]
> Backward should be accumulating the gradients on dxz. Here is the chunk of cuda code on what is happening. I have put comments in to explain what is happening.
> 
> The gradients of `dzx` can then be broken down into the `dx` and `dz` gradients with the following equation.
> 
> dx=dout⋅silu(z)
> 
> dz=dout⋅(z′)⋅x
> 
> Here z′ is the derivate of the silu function, but I was too lazy to put it in.
> 
> ```cuda-c++
> if constexpr (kHasZ) {
>            // We load in z out and dz
>             input_t *z = reinterpret_cast<input_t *>(params.z_ptr) + batch_id * params.z_batch_stride
>                 + dim_id * params.z_d_stride + chunk * kChunkSize;
>             input_t *out = reinterpret_cast<input_t *>(params.out_ptr) + batch_id * params.out_batch_stride
>                 + dim_id * params.out_d_stride + chunk * kChunkSize;
>             input_t *dz = reinterpret_cast<input_t *>(params.dz_ptr) + batch_id * params.dz_batch_stride
>                 + dim_id * params.dz_d_stride + chunk * kChunkSize;
>             input_t z_vals[kNItems], out_vals[kNItems];
>             __syncthreads();
>             load_input<Ktraits>(z, z_vals, smem_load, params.seqlen - chunk * kChunkSize);
>             __syncthreads();
>             load_input<Ktraits>(out, out_vals, smem_load, params.seqlen - chunk * kChunkSize);
>             float dz_vals[kNItems], z_silu_vals[kNItems];
>             // We calculate the gradients for each value described with the equations above
>             #pragma unroll
>             for (int i = 0; i < kNItems; ++i) {
>                 float z_val = z_vals[i];
>                 float z_sigmoid_val = 1.0f / (1.0f + expf(-z_val));
>                 z_silu_vals[i] = z_val * z_sigmoid_val;
>                 dz_vals[i] = dout_vals[i] * float(out_vals[i]) * z_sigmoid_val
>                              * (1.0f + z_val * (1.0f - z_sigmoid_val));
>                 dout_vals[i] *= z_silu_vals[i];
>             }
>             // Store store the output directly back into memory at dz
>             __syncthreads();
>             store_output<Ktraits>(dz, dz_vals, smem_store, params.seqlen - chunk * kChunkSize);
>             if (params.out_z_ptr != nullptr) {  // Recompute and store out_z
>                 float out_z_vals[kNItems];
>                 #pragma unroll
>                 for (int i = 0; i < kNItems; ++i) { out_z_vals[i] = float(out_vals[i]) * z_silu_vals[i]; }
>                 // if (blockIdx.x == 0 && blockIdx.y == 0 && threadIdx.x == 0) {
>                     // printf("out_val=%f, z_silu_val = %f, out_z_val = %f\n", float(out_vals[0]), z_silu_vals[0], out_z_vals[0]);
>                 // }
>                 input_t *out_z = reinterpret_cast<input_t *>(params.out_z_ptr) + batch_id * params.out_z_batch_stride
>                     + dim_id * params.out_z_d_stride + chunk * kChunkSize;
>                 __syncthreads();
>                 store_output<Ktraits>(out_z, out_z_vals, smem_store, params.seqlen - chunk * kChunkSize);
>             }
>         }
> ```
> 
> The storing will store the output directly in memory at that pointer. Since we first chunked `dz` and `dx` from `dzx` the memory pointers should still be pointing at the same tensor, i.e pytorch won't create two new tensors, but instead will point to two different positions in memory.
> 
> Thus it should update `dzx` with the values of `dx` and `dz`.
> 
> I am slightly confused about the testing though. The difference between those values is incredibly large, any clarification on that would be nice.

Thank you, very helpful explanation~
